### PR TITLE
refactor: prepare for later safety improvements

### PR DESCRIPTION
**Pull request by WalterBright**
I pulled out the common part of https://github.com/dlang/dmd/pull/5860 and made it a separate PR so that I can base more work on it without needing to pull all of 5860 just yet.

### CODE REVIEW

File: src/expression.d [hunk 1]
```diff
@@ -7824,6 +7824,18 @@ extern (C++) class UnaExp : Expression
...
 
+    /*********************
+     * Mark the operand as will never be dereferenced,
+     * which is useful info for @safe checks.
+     * Do before semantic() on operands rewrites them.
+     */
+    final void setNoderefOperand()
+    {
+        if (e1.op == TOKdotid)
+            (cast(DotIdExp)e1).noderef = true;
```
Comment (andralex) [inline src/expression.d]: space after `)` in cast throughout...

File: src/expression.d [hunk 2]
```diff
@@ -8484,6 +8510,7 @@ extern (C++) final class AssertExp : UnaExp
 extern (C++) final class DotIdExp : UnaExp
 {
     Identifier ident;
+    bool noderef;       // true if the result of the expression will never be dereferenced
```
Comment (andralex) [inline src/expression.d]: should be `noDeref`, I thought it meant "no dereference"

File: src/expression.d (hunk 1 above)
Comment (wilzbach) [inline src/expression.d]: Seems like we should enable Dscanner here too ;-)

File: src/mtype.d [hunk 3]
```diff
@@ -2705,13 +2705,13 @@ extern (C++) abstract class Type : RootObject
...
+                 * It should be gagged if flag & 1.
                  * e.g.
-                 *  tempalte opDispatch(name) if (isValid!name) { ... }
+                 *  template opDispatch(name) if (isValid!name) { ... }
                  */
-                uint errors = flag ? global.startGagging() : 0;
+                uint errors = flag & 1 ? global.startGagging() : 0;
                 e = dti.semanticY(sc, 0);
-                if (flag && global.endGagging(errors))
+                if (flag & 1 && global.endGagging(errors))
```
Comment (andralex) [inline src/mtype.d]: This subtly relies on the reader knowing the order of evaluation...

**Comment by andralex** [comment]
ehm

**Comment by JackStouffer** [comment]
Andrei, why do you always give advice on how to make a PR better but then merge it anyway?

**Comment by andralex** [comment]
Not always. At the end of the review I make the decision on whether that stops the merge or not.

**Pull request merged**